File: tests/difftok.c
// difftok: dump Runner's token ids for a corpus, for differential testing
// against the HuggingFace reference tokenizer.
//
//   difftok <model.gguf> <corpus.txt>
//
// The corpus is one JSON string literal per line (see
// scripts/tokenizer-corpus.py), so leading/trailing whitespace and newlines
// survive the file round trip -- those are exactly the cases that separate
// tokenizer implementations.
//
// Output is one line per corpus line: space-separated decimal token ids.
// Encoding runs with add_bos=false and parse_special=true. That is the exact
// counterpart of `tokenizers`' `encode(s, add_special_tokens=False)`: that flag
// only disables the post-processor (the BOS/EOS template), it does NOT stop
// added/special tokens in the *input text* from being matched. Passing
// parse_special=false here instead scores every "<s>"-shaped corpus string as a
// divergence, which is a harness artifact and not a tokenizer difference.
//
// BOS is left off on both sides: its placement is a chat-template concern, and
// folding it in would hide real merge differences behind a constant offset.
//
// scripts/difftok.py drives this and does the comparison.
#include "runner.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Decode one JSON string literal (including the surrounding quotes) in place.
// Returns the decoded length, or -1 if the line is not a JSON string.
static long unjson(const char *src, char *dst, size_t cap) {
    const char *p = src;
    while (*p == ' ' || *p == '\t') p++;
    if (*p != '"') return -1;
    p++;
    size_t n = 0;
    while (*p && *p != '"') {
        if (n + 4 >= cap) return -1;
        if (*p != '\\') { dst[n++] = *p++; continue; }
        p++;
        switch (*p) {
        case '"':  dst[n++] = '"';  p++; break;
        case '\\': dst[n++] = '\\'; p++; break;
        case '/':  dst[n++] = '/';  p++; break;
        case 'b':  dst[n++] = '\b'; p++; break;
        case 'f':  dst[n++] = '\f'; p++; break;
        case 'n':  dst[n++] = '\n'; p++; break;
        case 'r':  dst[n++] = '\r'; p++; break;
        case 't':  dst[n++] = '\t'; p++; break;
        case 'u': {
            p++;
            unsigned cp = 0;
            for (int i = 0; i < 4; i++) {
                char c = *p++;
                cp <<= 4;
                if (c >= '0' && c <= '9') cp |= (unsigned)(c - '0');
                else if (c >= 'a' && c <= 'f') cp |= (unsigned)(c - 'a' + 10);
                else if (c >= 'A' && c <= 'F') cp |= (unsigned)(c - 'A' + 10);
                else return -1;
            }
            if (cp >= 0xD800 && cp < 0xDC00 && p[0] == '\\' && p[1] == 'u') {
                unsigned lo = 0;
                const char *q = p + 2;
                for (int i = 0; i < 4; i++) {
                    char c = *q++;
                    lo <<= 4;
                    if (c >= '0' && c <= '9') lo |= (unsigned)(c - '0');
                    else if (c >= 'a' && c <= 'f') lo |= (unsigned)(c - 'a' + 10);
                    else if (c >= 'A' && c <= 'F') lo |= (unsigned)(c - 'A' + 10);
                    else return -1;
                }
                if (lo >= 0xDC00 && lo < 0xE000) {
                    cp = 0x10000 + ((cp - 0xD800) << 10) + (lo - 0xDC00);
                    p = q;
                }
            }
            if (cp < 0x80) dst[n++] = (char)cp;
            else if (cp < 0x800) {
                dst[n++] = (char)(0xC0 | (cp >> 6));
                dst[n++] = (char)(0x80 | (cp & 0x3F));
            } else if (cp < 0x10000) {
                dst[n++] = (char)(0xE0 | (cp >> 12));
                dst[n++] = (char)(0x80 | ((cp >> 6) & 0x3F));
                dst[n++] = (char)(0x80 | (cp & 0x3F));
            } else {
                dst[n++] = (char)(0xF0 | (cp >> 18));
                dst[n++] = (char)(0x80 | ((cp >> 12) & 0x3F));
                dst[n++] = (char)(0x80 | ((cp >> 6) & 0x3F));
                dst[n++] = (char)(0x80 | (cp & 0x3F));
            }
            break;
        }
        default: return -1;
        }
    }
    if (*p != '"') return -1;
    dst[n] = '\0';
    return (long)n;
}

int main(int argc, char **argv) {
    if (argc < 3) {
        fprintf(stderr, "usage: %s <model.gguf> <corpus.txt>\n", argv[0]);
        return 2;
    }

    gguf_file g;
    if (!gguf_open(&g, argv[1])) {
        fprintf(stderr, "difftok: cannot open %s\n", argv[1]);
        return 1;
    }
    tokenizer t;
    if (!tokenizer_init(&t, &g)) {
        fprintf(stderr, "difftok: cannot init tokenizer from %s\n", argv[1]);
        return 1;
    }

    FILE *f = fopen(argv[2], "rb");
    if (!f) {
        fprintf(stderr, "difftok: cannot open %s\n", argv[2]);
        return 1;
    }

    // Corpus strings are short by construction; the cap is generous.
    static char line[1 << 16];
    static char text[1 << 16];
    static int32_t ids[1 << 15];

    while (fgets(line, sizeof(line), f)) {
        size_t len = strlen(line);
        while (len && (line[len - 1] == '\n' || line[len - 1] == '\r'))
            line[--len] = '\0';
        if (len == 0) continue;

        if (unjson(line, text, sizeof(text)) < 0) {
            fprintf(stderr, "difftok: bad corpus line: %s\n", line);
            return 1;
        }
        int n = tok_encode(&t, text, ids, (int)(sizeof(ids) / sizeof(*ids)),
                           false, true);
        for (int i = 0; i < n; i++) printf("%s%d", i ? " " : "", ids[i]);
        printf("\n");
    }

    fclose(f);
    tokenizer_free(&t);
    gguf_close(&g);
    return 0;
}

File: scripts/difftok.py
#!/usr/bin/env python3
"""Tokenizer differential test: Runner vs the HuggingFace reference.

    scripts/difftok.py --gguf models/X.gguf --ref <hf-repo-id-or-dir> [--corpus F]

Builds the difftok harness (make difftok), runs it over the corpus, runs
`tokenizers` over the same corpus, and reports how many strings differ.

Both sides encode with no special tokens added: BOS/EOS placement is a chat
template concern, and folding it in here would mask real merge differences
behind a constant offset.

Divergences are printed with the decoded pieces from both sides, because "which
strings" matters far more than "how many" -- Mistral's known 2/721 is only
acceptable because every one of them begins with whitespace.

Exit status is 0 if the divergence count matches --expect (default 0).
"""

import argparse
import hashlib
import json
import os
import subprocess
import sys


def build(root, exe):
    make = os.environ.get("MAKE", "make")
    cc = os.environ.get("CC")
    cmd = [make, "-C", root, exe]
    if cc:
        cmd.append(f"CC={cc}")
    subprocess.run(cmd, check=True, stdout=subprocess.DEVNULL)


def load_corpus(path):
    out = []
    with open(path, encoding="utf-8") as f:
        for line in f:
            line = line.rstrip("\n").rstrip("\r")
            if not line:
                continue
            out.append(json.loads(line))
    return out


def runner_ids(exe, gguf, corpus_path, n):
    r = subprocess.run([exe, gguf, corpus_path], capture_output=True, text=True)
    if r.returncode != 0:
        sys.stderr.write(r.stderr)
        raise SystemExit(f"difftok harness failed ({r.returncode})")
    lines = r.stdout.split("\n")
    while lines and lines[-1] == "":
        lines.pop()
    if len(lines) != n:
        raise SystemExit(f"harness emitted {len(lines)} lines, corpus has {n}")
    return [[int(x) for x in ln.split()] for ln in lines]


def corpus_digest(corpus):
    """Identity of the exact strings a capture was taken against."""
    h = hashlib.sha256()
    for s in corpus:
        h.update(s.encode("utf-8"))
        h.update(b"\0")
    return h.hexdigest()


def reference_tokenizer(ref):
    from tokenizers import Tokenizer

    if os.path.isdir(ref):
        return Tokenizer.from_file(os.path.join(ref, "tokenizer.json"))
    if os.path.isfile(ref):
        return Tokenizer.from_file(ref)
    return Tokenizer.from_pretrained(ref)


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--gguf", required=True)
    ap.add_argument("--ref",
                    help="HF repo id, a directory containing tokenizer.json, "
                         "or a path to tokenizer.json. Not needed with "
                         "--ref-ids.")
    # Offline reference. Loading --ref by repo id needs the network and, for a
    # gated repo, credentials -- which is what keeps this check out of CI and
    # off contributors' machines. The reference tokenization of a fixed corpus
    # is a constant, though, so capture it ONCE from an authenticated run and
    # every later run is offline and reproducible.
    ap.add_argument("--capture", metavar="PATH",
                    help="write the reference ids for this corpus to PATH "
                         "(use during an authenticated run)")
    ap.add_argument("--ref-ids", metavar="PATH",
                    help="read reference ids from a --capture file instead of "
                         "loading a tokenizer; no network, no credentials")
    ap.add_argument("--corpus", default="tests/fixtures/tokenizer-corpus.txt")
    ap.add_argument("--expect", type=int, default=0,
                    help="expected number of diverging strings")
    ap.add_argument("--show", type=int, default=20,
                    help="max divergences to print in detail")
    ap.add_argument("--no-build", action="store_true")
    args = ap.parse_args()

    root = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    exe = os.path.join(root, "difftok.exe" if os.name == "nt" else "difftok")
    if not args.no_build:
        build(root, os.path.basename(exe))

    corpus = load_corpus(args.corpus)
    mine = runner_ids(exe, args.gguf, args.corpus, len(corpus))

    if args.ref_ids:
        with open(args.ref_ids, encoding="utf-8") as f:
            cap = json.load(f)
        # The ids mean nothing without the corpus they were captured against,
        # so refuse a mismatch loudly rather than comparing against the wrong
        # strings and reporting confident nonsense.
        want = corpus_digest(corpus)
        if cap.get("corpus_sha256") != want:
            raise SystemExit(
                "--ref-ids was captured against a different corpus "
                f"({cap.get('corpus_sha256', '?')[:12]} != {want[:12]}); "
                "recapture it")
        theirs = [list(x) for x in cap["ids"]]
        if len(theirs) != len(corpus):
            raise SystemExit("--ref-ids has %d rows, corpus has %d"
                             % (len(theirs), len(corpus)))
        tok = None
    else:
        if not args.ref:
            raise SystemExit("one of --ref or --ref-ids is required")
        tok = reference_tokenizer(args.ref)
        theirs = [tok.encode(s, add_special_tokens=False).ids for s in corpus]
        if args.capture:
            with open(args.capture, "w", encoding="utf-8") as f:
                json.dump({"ref": args.ref,
                           "corpus": os.path.basename(args.corpus),
                           "corpus_sha256": corpus_digest(corpus),
                           "ids": theirs}, f)
            print(f"captured {len(theirs)} reference tokenizations to "
                  f"{args.capture}")

    bad = [i for i in range(len(corpus)) if mine[i] != theirs[i]]

    for i in bad[: args.show]:
        s = corpus[i]
        print(f"--- line {i + 1}: {json.dumps(s, ensure_ascii=False)}")
        print(f"    runner: {mine[i]}")
        if tok is not None:
            print(f"            {[tok.decode([t]) if t < tok.get_vocab_size() else '?' for t in mine[i]]}")
        print(f"    ref   : {theirs[i]}")
        if tok is not None:
            print(f"            {[tok.decode([t]) for t in theirs[i]]}")
    if len(bad) > args.show:
        print(f"... and {len(bad) - args.show} more")

    leading_ws = sum(1 for i in bad if corpus[i][:1] in (" ", "\t", "\n", "\r"))
    print(f"\n{len(bad)}/{len(corpus)} strings differ "
          f"({leading_ws} of them begin with whitespace)")

    if len(bad) != args.expect:
        print(f"FAIL: expected {args.expect}", file=sys.stderr)
        return 1
    print("OK")
    return 0


if __name__ == "__main__":
    sys.exit(main())
